server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -364,7 +383,31 @@ public VolumeVO uploadVolume(UploadVolumeCmd cmd) throws ResourceAllocationExcep
         RegisterVolumePayload payload = new RegisterVolumePayload(cmd.getUrl(), cmd.getChecksum(), cmd.getFormat());
         vol.addPayload(payload);
 
-        volService.registerVolume(vol, store);
+        return registerVolume(vol, store, sizeInGB);
+    }
+
+    VolumeVO registerVolume(VolumeInfo vol, DataStore store, Long sizeInGB ) {
+        VolumeVO volume = _volsDao.findById(vol.getId());
+        try {
+            AsyncCallFuture<VolumeApiResult> future = volService.registerVolume(vol, store);
+            VolumeApiResult result = future.get();
+            if (result.isFailed()) {
+                s_logger.warn("Failed to register the volume");
+                String details = "";
+                if (result.getResult() != null && !result.getResult().isEmpty()) {
+                    details = result.getResult();
+                }
+                throw new CloudRuntimeException(details);
+            }
+            if (!Objects.equals(sizeInGB, vol.getSize())) {
+                if (sizeInGB !=null) {
+                    volume.setSize(sizeInGB * GiB_TO_BYTES);
```

Comment:
At this point is there any other reference that needs to be updated, like volume_store_ref? Could this cause any issues with GetVolumeStats or not considered at this volume state?